### PR TITLE
Integrate light emission

### PR DESCRIPTION
**Pull request by tobiaskleiner**
This draft serves to integrate the functionality to simulate light sources with the LightEmission package of sim_telarray within simtools.

There are two ways how the package can be run at the moment:
Example Usage
    -------------
1. Simulate light emission with varying distances:
      ```
       simtools-simulate-light-emission --telescope MSTN-04 --site North \
       --illuminator ILLN-01 --light_source_setup variable \
       --model_version prod6 --light_source_type led
      ```
 2. Simulate light emission with telescopes at fixed positions according to the layout:
       ```
       simtools-simulate-light-emission --telescope MSTN-04 --site North \
       --illuminator ILLN-01 --light_source_setup variable \
       --model_version prod6 --telescope_file \
       /workdir/external/simtools/tests/resources/telescope_positions-North-ground.ecsv\
       --light_source_type led
       ```
       
This creates output files in ```simtools/simtools-output/light_emission/``` and postscript files in ```simtools/simtools-output/light_emission/postscripts```. The postscripts can be then converted for instance with ps2pdf. (Not included in the container).

The coordinates of the illuminator (ILLN-01) are currently hardcoded in the configuration dictionaries. The telescope coordinates are obtained from the file given with ```--telescope_file```. This needs to be updated as soon as we have the coordinates in the database for the telescopes and calibration_devices.

The `xyzls` application used for the light-emission uses an isotropic light source.
Further updates are needed if we want to use the ls-beam application. A use case would be to configure a specific laser beam pointing, without changing the telescope pointings.

To make sure the telescope triggers, the array_trigger entry in the config is deleted.

The simtel_array output should contain something similar to:
```
Sim_telarray stage:
    Telescope 1 triggered (1/0/0/0, mask 1), summation from 36 to 95 of 105
    Event end data has been found.
    Shower of 0.0000 TeV energy was seen in 1 of 1 cases.
    Photon statistics:
    All photons:               928518
    Used photons:              928518
    Not absorbed/max. Q.E.:    189560
    Reflected on mirror:        26815
    Camera hit:                 25574
    Pixel hit:                  25574
    Detected:                   20998
    Trigger statistics:
    Tel. triggered:             1
    Tel. + array:               1
    Early readout:              0
    Late readout:               0
    Finish data conversion ...
    Writing 13 histograms to output file.
```
to indicate the telescope triggered. This means also successful writing of the `.simtel.gz` file. These files are used for postscript file creation and/ or pdf file creation with ctapipe.

### CODE REVIEW

**Comment by tobiaskleiner** [comment]
Running the example with two different configs:

1. % Configuration file for a NectarCam MST in Prod-3 (CTA-ULTRA6 setup). 
2. % ModelVersion: 2024-02-01 % TelescopeModelName: MSTN-design North

> Preprocessor command: /workdir/sim_telarray/sim_telarray/bin/pfp -v -I. -DNUM_TELESCOPES=1 -DWITH_LOW_GAIN_CHANNEL -DMAX_GAINS=2 -DWITH_GSL_RNG -DSIMTEL_VERSION=1709915244 -DSIMTEL_RELEASE=20240780 -I../cfg/CTAiobuf_maximum=1000000000 -I. -I/workdir/sim_telarray/sim_telarray/cfg -I/workdir/sim_telarray/sim_telarray/cfg/common -I/workdir/sim_telarray/sim_telarray/cfg/hess -I/workdir/sim_telarray/sim_telarray/cfg/hess2 -I/workdir/sim_telarray/sim_telarray/cfg/hess3 -I/workdir/sim_telarray/sim_telarray/cfg/hess5000 -I/workdir/sim_telarray/sim_telarray/cfg/CTA -I/workdir/sim_telarray/sim_telarray/cfg/MAGIC -DMAX_GAINS=2 -DTELESCOPE=4 - < /workdir/external/simtools/simtools-output/light_emission/model/CTA-North-MSTN-design-2024-02-01_light_emission.cfg
> Applying 0 weak command-line configuration parameter(s).
> Check additional configuration for telescope 4 from /workdir/external/simtools/simtools-output/light_emission/model/CTA-North-MSTN-design-2024-02-01_light_emission.cfg.
> Configuration for MSTN-design - TELESCOPE 4
> 132 configuration lines were processed.
> Applying 12 forced command-line configuration parameter(s).

Compraing the output of simtel_array `command += super()._config_option("show", "all")` below: (Not comparing camera settings as these are different analog_sum and majority) and ignoring lower part of configs (up to line 350 approx).

- `   Element name      Type    Elements  Lower b. Upper b.  Initial  Current`
- `    ARRAY_TRIGGERs    Text     4095                        none                           Current:   none` and `    ARRAY_TRIGGERs    Text     4095                        none                           Current:   array_trigger_prod6_lapalma_4_26_0.dat`
- `     IOBUF_OUTPUT_MAXimum                     Long     1        100000             20000000 400000000 `
- `    TRIGGER_CURRENT_LIMIT                       Double   1                           20.0     20 (2000)`
- `    TRIGGER_PIXELs    Int      1        1                  4        4  (3)`
- `    TRIGGER_TELESCOPES                       Int      1        1                  2        1 (2)` - set to 1 with -C now
- `    TELTRIG_MIN_TIME  Double   1        0.       10.       1.5      0.5 (1)`
- `    TELTRIG_MIN_SIGSUM                       Double   1        0.       1000.     7.8      7.8 (0)` - can't overwite with -C option


If I set the   -C option `command += super()._config_option("ARRAY_TRIGGERs", "None")` it gets ignored. If I remove the line in the config manually it gets updated before running the application. 

If I add a function within the light-emission code to remove the line, the telescope triggers, as previously when using the `remove_paramter`:

> Event 100 has triggered!
> Camera offset from nominal source = 0.00 cm, 0.00 cm
> E= 0.000 TeV, r=   0.00 m; Size= 2540.6 (193 pix.); width= 1.3903, length= 1.5293, dist= 0.1404, miss= 0.1365, alpha= 76.355 deg., Xmax=800.000, hmax=2241.71
> ... 
> Telescope 1 triggered (1/0/0/0, mask 1), summation from 41 to 100 of 128
> Event end data has been found.
> Shower of 0.0000 TeV energy was seen in 1 of 1 cases.
> Photon statistics:
>   All photons:                35136
>   Used photons:               35136
>   Not absorbed/max. Q.E.:     11085
>   Reflected on mirror:         1599
>   Camera hit:                  1567
>   Pixel hit:                   1567
>   Detected:                    1313
> Trigger statistics:
>   Tel. triggered:             1
>   Tel. + array:               1
>   Early readout:              0
>   Late readout:               0
> Finish data conversion ...
> Writing 13 histograms to output file.
> Closed output file.
> Finished.
> Writing histograms to file 'ctsim.hdata'.
> Histograms saved to 'ctsim.hdata'.
> 

However the output still looks different to previous results, and the trigger is not catching correctly the light emission. `TELTRIG_MIN_SIGSUM` set to 0 is correct?


The camera seems now to record mainly nsb (too large integration window etc.). Compare for instance the photon statistics for the light source at 4km distance: 

> Photon statistics:
>   All photons:                35136
>   Used photons:               35136
>   Not absorbed/max. Q.E.:     11085
>   Reflected on mirror:         1599
>   Camera hit:                  1567
>   Pixel hit:                   1567
>   Detected:                    1313
> Trigger statistics:
>   Tel. triggered:             1
>   Tel. + array:               1
>   Early readout:              0
>   Late readout:               0
> Finish data conversion ...

to the total p.e. in the camera :
total p. e. reco: 7352746.0

**Comment by GernotMaier** [comment]
> Running the example with two different configs:
> 
>     1. % Configuration file for a NectarCam MST in Prod-3 (CTA-ULTRA6 setup).

What does this mean? How and where do you set Prod3? This is very very outdated. 

> 
>     2. % ModelVersion: 2024-02-01 % TelescopeModelName: MSTN-design North
> 
> 
> > Preprocessor command: /workdir/sim_telarray/sim_telarray/bin/pfp -v -I. -DNUM_TELESCOPES=1 -DWITH_LOW_GAIN_CHANNEL -DMAX_GAINS=2 -DWITH_GSL_RNG -DSIMTEL_VERSION=1709915244 -DSIMTEL_RELEASE=20240780 -I../cfg/CTAiobuf_maximum=1000000000 -I. -I/workdir/sim_telarray/sim_telarray/cfg -I/workdir/sim_telarray/sim_telarray/cfg/common -I/workdir/sim_telarray/sim_telarray/cfg/hess -I/workdir/sim_telarray/sim_telarray/cfg/hess2 -I/workdir/sim_telarray/sim_telarray/cfg/hess3 -I/workdir/sim_telarray/sim_telarray/cfg/hess5000 -I/workdir/sim_telarray/sim_telarray/cfg/CTA -I/workdir/sim_telarray/sim_telarray/cfg/MAGIC -DMAX_GAINS=2 -DTELESCOPE=4 - < /workdir/external/simtools/simtools-output/light_emission/model/CTA-North-MSTN-design-2024-02-01_light_emission.cfg

File: simtools/simtel/simtel_light_emission.py
Comment (GernotMaier) [inline simtools/simtel/simtel_light_emission.py:359]: Why do you need an 'unused argument' statement here? Or remove the argument?

File: simtools/simtel/simtel_light_emission.py
Comment (GernotMaier) [inline simtools/simtel/simtel_light_emission.py:383]: Lot's of commented lines, do we want to remove them? It makes the code hard to read.

File: simtools/simtel/simtel_light_emission.py
Comment (GernotMaier) [inline simtools/simtel/simtel_light_emission.py:394]: Are these (and the following) parameters which should be configured? Or do they never change?

File: simtools/simtel/simtel_light_emission.py
Comment (GernotMaier) [inline simtools/simtel/simtel_light_emission.py:405]: I am not entirely sure about the coordinates (maybe the meaning of theta/phi should be documented in one of the docstrings). Are you sure that for both angles you need to do 180 - ...?

File: simtools/simtel/simtel_light_emission.py
Comment (GernotMaier) [inline simtools/simtel/simtel_light_emission.py:413]: Surprised to see that a power law is needed. This is used for what?

File: simtools/simtel/simtel_light_emission.py
Comment (GernotMaier) [inline simtools/simtel/simtel_light_emission.py:425]: Same as above, why not remove the argument?

File: simtools/simtel/simtel_light_emission.py
Comment (GernotMaier) [inline simtools/simtel/simtel_light_emission.py:434]: The integration parameters are the same for all telescopes? Remember we have different readout speeds.

File: simtools/simtel/simtel_light_emission.py
Comment (GernotMaier) [inline simtools/simtel/simtel_light_emission.py:476]: Hardwired cleaning parameters - this is / should be a configuration parameter?

**Review by GernotMaier — CHANGES_REQUESTED** [review]
Thanks for this - very nice!

A couple of comments in the the review part (also some related to using it; I am not entirely familiar with every details).

- important is to separate the simulation from the plotting
- a couple of commented lines which probably should go at some point
- not sure about the testing: for unit tests, we usually try to have for every function a dedicated test function to test the function 'as a unit'. The tests seems to be more  related to the full module / application?

File: simtools/db/db_handler.py
Comment (tobiaskleiner) [reply simtools/db/db_handler.py]: Removed. I added this as an intermediate fix but forgot to delet it later.

File: simtools/db/db_handler.py
Comment (tobiaskleiner) [reply simtools/db/db_handler.py]: Changed back to use default collection.

File: simtools/db/db_handler.py
Comment (tobiaskleiner) [reply simtools/db/db_handler.py]: Changed.

File: simtools/simtel/simtel_light_emission.py
Comment (tobiaskleiner) [reply simtools/simtel/simtel_light_emission.py]: Done

File: simtools/simtel/simtel_light_emission.py
Comment (tobiaskleiner) [reply simtools/simtel/simtel_light_emission.py:458]: I have added the command line parameters.

File: simtools/simtel/simtel_light_emission.py
Comment (tobiaskleiner) [reply simtools/simtel/simtel_light_emission.py]: Added the command line parameters.

File: simtools/simtel/simtel_light_emission.py
Comment (tobiaskleiner) [reply simtools/simtel/simtel_light_emission.py:458]: Ok I can also add these to the command line.

File: simtools/simtel/simtel_light_emission.py
Comment (tobiaskleiner) [reply simtools/simtel/simtel_light_emission.py:432]: Also not really sure if needed, but this appeared in the examples by Konrad, so there must be a reason for it probably. No info in the light emission docs about it.

File: simtools/db/db_handler.py
Comment (tobiaskleiner) [reply simtools/db/db_handler.py]: Changed.

File: simtools/layout/array_layout.py
Comment (tobiaskleiner) [reply simtools/layout/array_layout.py:441]: Ok good I see.

File: simtools/applications/light_emission.py
Comment (tobiaskleiner) [reply simtools/applications/light_emission.py]: Fixed

File: simtools/applications/light_emission.py
Comment (tobiaskleiner) [reply simtools/applications/light_emission.py]: Fixed

File: simtools/applications/light_emission.py
Comment (tobiaskleiner) [reply simtools/applications/light_emission.py]: Fixed

File: simtools/applications/light_emission.py
Comment (tobiaskleiner) [reply simtools/applications/light_emission.py:235]: I have removed most of them now. The remaining ones are for the coordinates and can be removed as well later when we have the coordinates in the DB.

File: simtools/applications/light_emission.py
Comment (tobiaskleiner) [reply simtools/applications/light_emission.py]: Removed

File: simtools/applications/light_emission.py
Comment (tobiaskleiner) [reply simtools/applications/light_emission.py:383]: Yes this was my thought. If not needed I can remove this.

File: simtools/model/calibration_model.py
Comment (tobiaskleiner) [reply simtools/model/calibration_model.py]: Adjusted

File: tests/unit_tests/simtel/test_simtel_light_emission.py
Comment (tobiaskleiner) [reply tests/unit_tests/simtel/test_simtel_light_emission.py]: Removed

File: tests/unit_tests/simtel/test_simtel_light_emission.py
Comment (tobiaskleiner) [reply tests/unit_tests/simtel/test_simtel_light_emission.py]: Added this now. I wonder if this triggered the pylint issues that we see now in the CI.

File: tests/unit_tests/simtel/test_simtel_light_emission.py
Comment (tobiaskleiner) [reply tests/unit_tests/simtel/test_simtel_light_emission.py]: Fixed.

File: tests/unit_tests/simtel/test_simtel_light_emission.py
Comment (tobiaskleiner) [reply tests/unit_tests/simtel/test_simtel_light_emission.py]: Removed.

File: tests/integration_tests/config/simulate_light_emission_layout.yml
Comment (tobiaskleiner) [reply tests/integration_tests/config/simulate_light_emission_layout.yml:1]: No this is not expected. I have just tried and I can run both ways. Was this the layout or variable commands for the light source?

File: simtools/simtel/simtel_light_emission.py
Comment (tobiaskleiner) [reply simtools/simtel/simtel_light_emission.py]: Ok changed this.

File: simtools/simtel/simtel_light_emission.py
Comment (tobiaskleiner) [reply simtools/simtel/simtel_light_emission.py]: Removed.

File: simtools/simtel/simtel_light_emission.py
Comment (tobiaskleiner) [reply simtools/simtel/simtel_light_emission.py]: Fixed.

File: simtools/simtel/simtel_light_emission.py
Comment (tobiaskleiner) [reply simtools/simtel/simtel_light_emission.py:327]: You mean to copy them in the include directory in the simtel_array include directory? I believe the better way would be to change the application to accept a path. Otherwise this is difficult to keep track of if and what file was actually used in the end.

File: simtools/simtel/simtel_light_emission.py
Comment (tobiaskleiner) [reply simtools/simtel/simtel_light_emission.py]: Removed.

File: simtools/simtel/simtel_light_emission.py
Comment (tobiaskleiner) [reply simtools/simtel/simtel_light_emission.py:404]: Yes this is true. I could remove them.

File: simtools/db/db_handler.py
Comment (tobiaskleiner) [reply simtools/db/db_handler.py:1]: Do you mean in general or in the case of calibration_device? If you want I can add another issue for that?

Comment (GernotMaier) [reply simtools/db/db_handler.py:1]: Very general, but let me open an issue on that. I think this is beyond this pull request.

File: ctsim.hdata
Comment (GernotMaier) [inline ctsim.hdata:1]: Die Datei sollte aber nicht da sein, oder?

File: simtools/applications/light_emission.py
Comment (GernotMaier) [reply simtools/applications/light_emission.py:164]: It is not clear that these are two different applications. I suggest that you add some additional documentation into the docstr.

File: simtools/db/db_handler.py
Comment (GernotMaier) [reply simtools/db/db_handler.py:1]: #918

File: simtools/simtel/simtel_light_emission.py
Comment (GernotMaier) [reply simtools/simtel/simtel_light_emission.py:242]: OK. But please open an issue.

File: simtools/simtel/simtel_light_emission.py
Comment (GernotMaier) [reply simtools/simtel/simtel_light_emission.py:327]: No, simtools would pull the profiles from the DB for each run, so this way we can be sure to use the right ones.

File: simtools/simtel/simtel_light_emission.py
Comment (GernotMaier) [reply simtools/simtel/simtel_light_emission.py:360]: Please add this as comment .

File: simtools/simtel/simtel_light_emission.py
Comment (GernotMaier) [reply simtools/simtel/simtel_light_emission.py:415]: Did you see this question?

**Review by GernotMaier — APPROVED** [review]
Thanks for this - see below for some minor comments and changes requested.

I approve already, so after adding the improved docstrings, please merge.

File: ctsim.hdata
Comment (tobiaskleiner) [reply ctsim.hdata:1]: Removed

File: tests/integration_tests/test_applications_from_config.py
Comment (GernotMaier) [inline tests/integration_tests/test_applications_from_config.py:355]: Final request (last one) - change this to:

f"Model version {model_version_requested) not supported for this this"

**Review by GernotMaier — APPROVED** [review]
Super minor request - then good!

**Comment by tobiaskleiner** [comment]
Thanks Gernot for the review and the comments. I have adressed the remaining ones. Going to merge then as soon as the tests have passed.

**Pull request merged**
